server/main_server/src/agent_main.cpp

```diff
@@ -0,0 +1,1326 @@
+                time_out.tv_usec = 500 * 1000;
+
+                if (EINTR == errno) {
+                    log_af::trace("{}: select() interrupted", __func__);
+                    // NOLINTNEXTLINE(hicpp-signed-bitwise, cppcoreguidelines-pro-bounds-constant-array-index)
+                    FD_SET(svrComm.sock, &sockMask);
+                    continue;
+                }
+
+                log_af::error("{}: select() error, errno = {}", __func__, errno);
+                return 1;
+            }
+
+            if (g_terminate) {
+                log_af::info("{}: Received shutdown instruction. Exiting agent factory main loop.", __func__);
+                break;
+            }
+
+            if (g_terminate_graceful) {
+                log_af::info("{}: Received shutdown instruction. Exiting agent factory main loop.", __func__);
+                break;
+            }
+
+            if (0 == numSock) {
+                // "select" modifies the timeval structure, so reset it.
+                time_out = original_time_out;
+                continue;
+            }
+
+            const int newSock = rsAcceptConn(&svrComm);
+            if (newSock < 0) {
+                log_af::info("{}: rsAcceptConn() error, errno = {}", __func__, errno);
+                continue;
+            }
+
+            // Fork agent to handle client request.
+
+            const auto agent_pid = fork();
+            if (agent_pid == 0) {
+                close(svrComm.sock); // Close the listening socket.
+                try {
+                    return handle_client_request(newSock, std::time(nullptr));
+                }
+                catch (const std::exception& e) {
+                    log_agent::critical("{}: Exception caught in outer most scope of request handler: {}", __func__, e.what());
+                    return 1;
+                }
+            }
+
+            if (agent_pid > 0) {
+                close(newSock);
+            }
+            else {
+                log_af::critical("Failed to fork agent. errno={}", errno);
+            }
+        }
+
+        // Start shutting everything down.
+
+        // Do not accept new client requests (i.e. close the listening socket).
+        close(svrComm.sock);
+
+        // Give the rule engine plugins a chance to free any resources established during setup().
+        irods::re_plugin_globals->global_re_mgr.call_teardown_operations();
+
+        if (0 == g_terminate_graceful) {
+            // Instruct all agents to shutdown gracefully.
+            // To avoid unnecessary complexity, we use SIGUSR1 as the termination signal for agents.
+            // Attempting to use SIGTERM would also notify the main server process and delay server.
+            kill(0, SIGUSR1);
+        }
+
+        reap_agent_processes(true);
+
+        log_af::info("{}: Shutdown complete.", __func__);
+
+        return 0;
+    }
+    catch (const std::exception& e) {
+        fmt::print(stderr, "Error: {}\n", e.what());
```

Comment:
Verify if this should use the logger instead of fmtlib. I think the logger isn't guaranteed to be initialized at this level.

Reply:
The code highlighted is correct.